avm/src/vm/memory.rs

```diff
@@ -136,67 +149,132 @@ impl HandlerMemory {
     }
   }
 
-  /// Takes a given address and looks up the `mems` indexes relevant to it.
-  fn addr_to_idxs(self: &HandlerMemory, addr: i64) -> (usize, usize) {
+  /// Drop the immutable, thread-safe reference to the parent provided by `fork`.
+  /// Consuming forked children is a mutable operation on the parent
+  /// that cannot be performed on an object within an Arc. As a result, 
+  /// all the forked children drop their reference to the parent, we take
+  /// the parent out of the Arc after parallel work is done and then join on its children.
+  pub fn drop_parent(self: &mut HandlerMemory) {
+    if self.parent.is_some() {
+      let arc = self.parent.take();
+      drop(arc);
+    }
+  }
+
+  /// Returns true if the idxs are valid in self
+  fn is_idx_defined(self: &HandlerMemory, a: usize, b: usize) -> bool {
+    let is_raw = a == std::usize::MAX;
+    let safe_mem_space = self.mem_addr == 1 || a >= self.mem_addr;
+    let is_fixed = safe_mem_space && self.mems.len() > a && self.mems[a].len() > b;
+    let is_fractal = safe_mem_space && self.mems.len() > a && b == std::usize::MAX;
+    return is_raw || is_fixed || is_fractal;
+  }
+
+  /// Recursively finds the parent and returns None if the idxs
+  /// belong to self or otherwise a reference to the parent
+  fn hm_for_idxs(self: &HandlerMemory, a: usize, b: usize) -> Option<&Arc<HandlerMemory>> {
+    if self.is_idx_defined(a,b) {
+      return None;
+    }
+    let res = self.parent.as_ref().unwrap().hm_for_idxs(a, b);
+    return if res.is_none() { self.parent.as_ref() } else { res };
+  }
+
+  /// Takes a given address and looks up the fractal location and
+  /// `mems` indexes relevant to it if available
+  fn addr_to_idxs_opt(self: &HandlerMemory, addr: i64) -> Option<(usize, usize)> {
     return if addr >= 0 {
-     self.addr.0[addr as usize]
+      *self.addr.0.get(addr as usize).unwrap_or(&None)
```

Comment:
I'm really surprised this syntax works. If the address is defined, it'll unwrap it and return a pointer to it, but that somehow gets automatically turned into a `Some` Option? If it fails it's a pointer to the address of `None`, which kinda makes sense but still seems super indirect.

Reply:
Discussed offline, the `get` method called for `Vec` returns a `&Option<(usize, usize)>` versus before it was just `(usize, usize)`

Reply:
It returns an `Option<&Option<(usize, usize)>>` to be precise. :)